### ISSUE TITLE
Eager execution support in tf.keras fit_generator breaks Keras callbacks

### ISSUE THREAD

**Issue by smatzek** [tensorflow/tensorflow#21162]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: RHEL 7.5 ppc64le but would occur on others
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.8
- **Python version**: Anaconda3, Python 3.6.4
- **Bazel version (if compiling from source)**:  N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: 9.2
- **GPU model and memory**: NVIDIA Volta V100
- **Exact command to reproduce**: complicated, see below

### Describe the problem
When using tf.keras fit_generator with Keras callbacks the backward phase of the model is not present when the set_model, set_params, and on_train_begin callback methods are called.  This breaks any callback what needs the gradients, ops, etc present in the model before training begins.

This change was injected by this commit for eager execution support:
https://github.com/tensorflow/tensorflow/commit/1b67ccbe8006eacffd268553abd01310e8b187d6#diff-abfd39a5a5f655f8d92c482e91081f8c

The _make_train_function that was removed calls the optimizer's get_updates() method which adds all the operations under the training/<optimizer class name>/ scope. 

Once possible solution to this may be to put the calls back in but guard them with `if context.executing_eagerly()` as is done here:
https://github.com/tensorflow/tensorflow/blob/939237af1479cf61a7c16ad4a1ead521da65a3a6/tensorflow/python/keras/engine/training.py#L1640-L1650

### Source code / logs
An example of a Keras callback that depends on the full model being present when the set_model function is called is this model graph analysis / rewriting callback:
https://github.com/tensorflow/tensorflow/pull/19845/files#diff-ae5805990d1ab77911507c9ae0b1cda2

**Comment by smatzek** [comment]
@robieta, I didn't know when I'd get a chance to write and code a test for this so I didn't self assign.  I've since had a chance to test out the patch with both my callback and the original eager example that was the cause for the removal.  I've put up PR:
https://github.com/tensorflow/tensorflow/pull/21244

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by smatzek** [comment]
This issue was fixed under #21244
